book/src/storage/storage-functions.md

```diff
@@ -0,0 +1,242 @@
+
+> Function definitions are not connected to the object's state. It is possible to define a function that takes a mutable reference to an object that is used as frozen. However, it won't be callable on a frozen object.
+
+The `message` function can be called on an immutable `Config` object, however, two functions below are not callable on a frozen object:
+
+```move
+// === Functions below can't be called on a frozen object! ===
+
+/// The function can be defined, but it won't be callable on a frozen object.
+/// Only immutable references are allowed.
+public fun message_mut(c: &mut Config): &mut String { &mut c.message }
+
+/// Deletes the `Config` object, takes it by value.
+/// Can't be called on a frozen object!
+public fun delete_config(c: Config) {
+    let Config { id, message: _ } = c;
+    id.delete()
+}
+```
+
+To summarize:
+- `freeze_object` function is used to put an object into an _immutable_ state;
+- Once an object is _frozen_, it can never be changed, deleted or transferred, and it can be accessed by anyone by immutable reference;
+
+## Owned -> Frozen
+
+Since the `transfer::freeze` signature accepts any type with the `key` ability, it can take an object that was created in the same scope, but it can also take an object that was owned by an account. This means that the `freeze_object` function can be used to _freeze_ an object that was _transferred_ to the sender. For security concerns, we would not want to freeze the `AdminCap` object - it would be a security risk to allow access to it to anyone. However, we can freeze the `Gift` object that was minted and transferred to the recipient:
+
+> Single Owner -> Immutable conversion is possible!
+
+```move
+/// Freezes the `Gift` object so it becomes immutable.
+public fun freeze_gift(gift: Gift) {
+    transfer::freeze_object(gift);
+}
+```
+
+## Share
+
+The `transfer::share` function is a public function used to put an object into a _shared_ state. Once an object is _shared_, it can be accessed by anyone by a mutable reference (hence, immutable too). The function signature is as follows, only accepts a type with the [`key` ability](./key-ability.md):
+
+```move
+// File: sui-framework/sources/transfer.move
+public fun share_object<T: key>(obj: T);
+```
+
+Once an object is _shared_, it is publicly available as a mutable reference.
+
+## Special Case: Shared Object Deletion
+
+While the shared object can't normally be taken by value, there is one special case where it can - if the function that takes it deletes the object. This is a special case in the Sui storage model, and it is used to allow the deletion of shared objects. To show how it works, we will create a function that creates and shares a Config object and then another one that deletes it:
```

Comment:
The rule is a bit more relaxed than this. A shared object can be taken by value, but it has to be reshared or deleted by the end of the transaction. This means you could thread the object around by value. 
cc @tzakian to sanity check me here